zenml/conftest.py

```diff
@@ -0,0 +1,241 @@
+        try:
+            local_store = os.path.join(ZENML_ROOT, local_store_path)
+            shutil.rmtree(local_store, ignore_errors=False)
+        except Exception as e:
+            print(e)
+
+    return wrapper
+
+
+@pytest.fixture
+def cleanup(cleanup_metadata_store, cleanup_artifacts):
+    cleanup_metadata_store()
+    cleanup_artifacts()
+
+
+@pytest.fixture
+def run_test_pipelines():
+    def wrapper():
+        repo: Repository = Repository.get_instance()
+        repo.zenml_config.set_pipelines_dir(pipeline_root)
+
+        for p_config in path_utils.list_dir(pipeline_root):
+            y = yaml_utils.read_yaml(p_config)
+            p: TrainingPipeline = TrainingPipeline.from_config(y)
+            p.run()
+    return wrapper
+
+
+@pytest.fixture
+def delete_config():
+    def wrapper(filename):
+        repo: Repository = Repository.get_instance()
+        repo.zenml_config.set_pipelines_dir(pipeline_root)
+
+        try:
+            cfg = os.path.join(pipeline_root, filename)
+            os.remove(cfg)
+        except Exception as e:
+            print(e)
+            pass
+
+    return wrapper
+
+
+@pytest.fixture
+def equal_backends():
+    def wrapper(bck1: BaseBackend, bck2: BaseBackend, loaded=True):
+        # There can be a "None" backend in a pipeline
+        if bck1 is None and bck2 is None:
+            return True
+        if sum(b is None for b in [bck1, bck2]) == 1:
+            return False
+
+        equal = False
+        equal |= bck1._kwargs == bck2._kwargs
+        equal |= bck1._source == bck2._source
+        if loaded:
+            equal |= bck1._immutable != bck2._immutable
+        else:
+            equal |= bck1._immutable == bck2._immutable
+
+        return equal
+
+    return wrapper
+
+
+@pytest.fixture
+def equal_steps(equal_backends):
+    def wrapper(step1: BaseStep, step2: BaseStep, loaded=True):
+        # There can be a "None" step e.g. in get_data_step()
+        # in a BaseDatasource
+        if step1 is None and step2 is None:
+            return True
+        if sum(s is None for s in [step1, step2]) == 1:
+            return False
+
+        equal = False
+        equal |= step1._kwargs == step2._kwargs
+        equal |= equal_backends(step1.backend, step2.backend, loaded=loaded)
+        equal |= step1._source == step2._source
+        if loaded:
+            equal |= step1._immutable != step2._immutable
+        else:
+            equal |= step1._immutable == step2._immutable
+
+        return equal
+
+    return wrapper
+
+
+@pytest.fixture
+def equal_datasources(equal_steps):
+    def wrapper(ds1: BaseDatasource, ds2: BaseDatasource, loaded=True):
+        # There can be a "None" datasource in a pipeline
+        if ds1 is None and ds2 is None:
+            return True
+        if sum(d is None for d in [ds1, ds2]) == 1:
+            return False
+
+        equal = False
+        equal |= ds1.name == ds2.name
+        equal |= ds1.schema == ds2.schema
```

Comment:
Will this work, as these are both dicts

Reply:
It does: 

https://stackoverflow.com/questions/17217225/what-does-the-operator-actually-do-on-a-python-dictionary

https://stackoverflow.com/questions/58961051/python-3-6-equality-of-two-dictionaries-with-same-keys-but-in-different-order

Also, it does so independently of order, so we do not need to implement any more elaborate equality tester.